File: API/1562_XAU_USD_with_Correct_ADX_and_Bollinger_Bands_Fill/CS/XauUsdAdxBollingerStrategy.cs
using System;
using System.Linq;
using System.Collections.Generic;

using Ecng.Common;
using Ecng.Collections;
using Ecng.Serialization;

using StockSharp.Algo.Indicators;
using StockSharp.Algo.Strategies;
using StockSharp.BusinessEntities;
using StockSharp.Messages;

namespace StockSharp.Samples.Strategies;

/// <summary>
/// XAU/USD strategy using Bollinger Bands breakout with trend strength filter.
/// Uses SMA + StdDev as Bollinger proxy. Trades band breakouts when trend is strong.
/// </summary>
public class XauUsdAdxBollingerStrategy : Strategy
{
	private readonly StrategyParam<int> _bollPeriod;
	private readonly StrategyParam<decimal> _bbWidth;
	private readonly StrategyParam<int> _trendLength;
	private readonly StrategyParam<DataType> _candleType;

	private readonly List<decimal> _closes = new();

	public int BollingerPeriod { get => _bollPeriod.Value; set => _bollPeriod.Value = value; }
	public decimal BbWidth { get => _bbWidth.Value; set => _bbWidth.Value = value; }
	public int TrendLength { get => _trendLength.Value; set => _trendLength.Value = value; }
	public DataType CandleType { get => _candleType.Value; set => _candleType.Value = value; }

	public XauUsdAdxBollingerStrategy()
	{
		_bollPeriod = Param(nameof(BollingerPeriod), 20)
			.SetGreaterThanZero()
			.SetDisplay("Bollinger Period", "SMA/StdDev period", "Indicators");

		_bbWidth = Param(nameof(BbWidth), 2m)
			.SetGreaterThanZero()
			.SetDisplay("BB Width", "Band multiplier", "Indicators");

		_trendLength = Param(nameof(TrendLength), 14)
			.SetGreaterThanZero()
			.SetDisplay("Trend Length", "Directional movement lookback", "Indicators");

		_candleType = Param(nameof(CandleType), TimeSpan.FromMinutes(5).TimeFrame())
			.SetDisplay("Candle Type", "Type of candles", "General");
	}

	public override IEnumerable<(Security sec, DataType dt)> GetWorkingSecurities()
	{
		return [(Security, CandleType)];
	}

	protected override void OnReseted()
	{
		base.OnReseted();
		_closes.Clear();
	}

	protected override void OnStarted2(DateTime time)
	{
		base.OnStarted2(time);

		var sma = new SimpleMovingAverage { Length = BollingerPeriod };
		var stdDev = new StandardDeviation { Length = BollingerPeriod };

		_closes.Clear();

		var subscription = SubscribeCandles(CandleType);
		subscription.Bind(sma, stdDev, ProcessCandle).Start();

		var area = CreateChartArea();
		if (area != null)
		{
			DrawCandles(area, subscription);
			DrawIndicator(area, sma);
			DrawOwnTrades(area);
		}
	}

	private void ProcessCandle(ICandleMessage candle, decimal smaVal, decimal stdVal)
	{
		if (candle.State != CandleStates.Finished)
			return;

		_closes.Add(candle.ClosePrice);
		while (_closes.Count > TrendLength + 1)
			_closes.RemoveAt(0);

		if (stdVal <= 0 || _closes.Count < TrendLength)
			return;

		var upper = smaVal + BbWidth * stdVal;
		var lower = smaVal - BbWidth * stdVal;

		// Simple trend strength: absolute price change over period / average range
		var priceChange = Math.Abs(candle.ClosePrice - _closes[0]);
		var avgChange = stdVal * 2; // approximate
		var trendStrength = avgChange > 0 ? priceChange / avgChange : 0;

		// Only trade when trend is reasonably strong
		if (trendStrength > 0.5m)
		{
			if (candle.ClosePrice > upper && Position <= 0)
				BuyMarket();
			else if (candle.ClosePrice < lower && Position >= 0)
				SellMarket();
		}

		// Mean reversion exit
		if (Position > 0 && candle.ClosePrice < smaVal)
			SellMarket();
		else if (Position < 0 && candle.ClosePrice > smaVal)
			BuyMarket();
	}
}

File: API/1562_XAU_USD_with_Correct_ADX_and_Bollinger_Bands_Fill/PY/xau_usd_adx_bollinger_strategy.py
import clr

clr.AddReference("StockSharp.Messages")
clr.AddReference("StockSharp.Algo")

from System import TimeSpan, Math
from StockSharp.Messages import DataType, CandleStates
from StockSharp.Algo.Indicators import SimpleMovingAverage, StandardDeviation
from StockSharp.Algo.Strategies import Strategy


class xau_usd_adx_bollinger_strategy(Strategy):
    def __init__(self):
        super(xau_usd_adx_bollinger_strategy, self).__init__()
        self._boll_period = self.Param("BollingerPeriod", 20) \
            .SetDisplay("Bollinger Period", "SMA/StdDev period", "Indicators")
        self._bb_width = self.Param("BbWidth", 2.0) \
            .SetDisplay("BB Width", "Band multiplier", "Indicators")
        self._trend_length = self.Param("TrendLength", 14) \
            .SetDisplay("Trend Length", "Directional movement lookback", "Indicators")
        self._candle_type = self.Param("CandleType", DataType.TimeFrame(TimeSpan.FromHours(4))) \
            .SetDisplay("Candle Type", "Type of candles", "General")
        self._closes = []

    @property
    def boll_period(self):
        return self._boll_period.Value

    @property
    def bb_width(self):
        return self._bb_width.Value

    @property
    def trend_length(self):
        return self._trend_length.Value

    @property
    def candle_type(self):
        return self._candle_type.Value

    def OnReseted(self):
        super(xau_usd_adx_bollinger_strategy, self).OnReseted()
        self._closes = []

    def OnStarted(self, time):
        super(xau_usd_adx_bollinger_strategy, self).OnStarted(time)
        sma = SimpleMovingAverage()
        sma.Length = self.boll_period
        std_dev = StandardDeviation()
        std_dev.Length = self.boll_period
        subscription = self.SubscribeCandles(self.candle_type)
        subscription.Bind(sma, std_dev, self.on_process).Start()
        area = self.CreateChartArea()
        if area is not None:
            self.DrawCandles(area, subscription)
            self.DrawIndicator(area, sma)
            self.DrawOwnTrades(area)

    def on_process(self, candle, sma_val, std_val):
        if candle.State != CandleStates.Finished:
            return
        sma_val = float(sma_val)
        std_val = float(std_val)
        close = float(candle.ClosePrice)
        self._closes.append(close)
        while len(self._closes) > int(self.trend_length) + 1:
            self._closes.pop(0)
        if std_val <= 0 or len(self._closes) < int(self.trend_length):
            return
        bb_w = float(self.bb_width)
        upper = sma_val + bb_w * std_val
        lower = sma_val - bb_w * std_val
        price_change = abs(close - self._closes[0])
        avg_change = std_val * 2.0  # approximate
        trend_strength = price_change / avg_change if avg_change > 0 else 0
        if trend_strength > 1:
            if close > upper and self.Position <= 0:
                self.BuyMarket()
            elif close < lower and self.Position >= 0:
                self.SellMarket()

    def CreateClone(self):
        return xau_usd_adx_bollinger_strategy()
